Enable hipBLASLt backend for GEMM, and make it the default option for RNN's under specific conditions

BrianHarrisonAMD opened the pull request:
Changes:
- Update MIOpen to support hipBLASLt backend for Gemm via an override Environment variable.

Known remaining work:
- Add additional workspace calculation to RNNs to account for extra memory needed by hipBLASLt for optimal algorithm selection. 
    - Note: this will be done with future work to enable hipBLASLt fully, and convert to use solvers.

BrianHarrisonAMD committed:
Add hipBLASLt GEMM backend option and enable for RNN under specific conditions

BrianHarrisonAMD committed:
Move HipEventProfiler to cpp, and fix warnings

BrianHarrisonAMD committed:
Apply formatting

shurale-nkn commented on include/miopen/config.h.in:
```diff
@@ -108,7 +108,7 @@ const char* getOffloadBundlerBinPath();
 #cmakedefine01 MIOPEN_USE_SQLITE_PERFDB
 
-#define MIOPEN_USE_GEMM (MIOPEN_USE_ROCBLAS)
+#define MIOPEN_USE_GEMM (MIOPEN_USE_ROCBLAS || MIOPEN_BACKEND_HIP)
```
MIOPEN_BACKEND_HIP? is it guaranteed that the component is always installed in the system with HIP?

BrianHarrisonAMD replied:
When I spoke with Jehandad about this task, he said that was a safe assumption.

I believe we might even be able to assume HIP as well now, and remove the define checks entirely.

BrianHarrisonAMD replied:
Started a discussion, and the result is that we are going to assume HIP is installed / a dependency, and hipBLASLt will be installed / a dependency in all cases as well.

CAHEK7 commented on src/gemm_v2.cpp:
```diff
@@ -324,6 +326,173 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+        }                                                \
+    } while(0)
+
+template <typename DataTypeAB, hipDataType HipTypeAB, typename DataTypeC, hipDataType HipTypeC>
```
I don't think that `HipTypeAB` should be a template parameter. It bloats the binary instantiating unique function for each distinct type.

CAHEK7 replied:
I don't think that HipTypeAB should be a template parameter. It bloats the binary instantiating unique function for each distinct type.

CAHEK7 commented on src/gemm_v2.cpp:
```diff
@@ -324,6 +326,173 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+        {                                                \
+            return e;                                    \
+        }                                                \
+    } while(0)
```
I would throw an exception and make it a function instead of define.
Moreover the exception has already been thrown, but for no reasons it happened few layer later, at line 802, 1096 and 1388. Those checks and the status propagating can be removed as well.

BrianHarrisonAMD replied:
Thoughts on the updated version @CAHEK7?

I made the changes, but still required a macro to make it easy to call the templated method that throws when status isn't success. I think this was the cleanest way to do this, but curious if you had a suggestion that would be better.

Thanks!

CAHEK7 replied:
It can be something like this:
```C++
static inline void hipblas_status_check(hipblasStatus_t status)
{
    if(status != hipblasStatus_t::HIPBLAS_STATUS_SUCCESS)
    {
        MIOPEN_THROW(miopenStatusInternalError, "hipBlasLt error encountered");
    }
}

hipblas_status_check(gemm.setProblem(...));
```
There is no need to pass whole function into the checker, we need just a status.

BrianHarrisonAMD replied:
hehe good point! 

I will update. Thanks!

CAHEK7 commented on src/gemm_v2.cpp:
```diff
@@ -324,6 +326,173 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+    }
+    break;
+    case miopenFloat8: {
+        const auto is_gfx94x = miopen::StartsWith(handle.GetDeviceName(), "gfx94");
```
It is not related to this particular PR, but we shall add proper capabilities check into the `handle`, those string comparisons are ugly, slow and, the most important, error prone.

CAHEK7 commented on src/gemm_v2.cpp:
```diff
@@ -324,6 +326,173 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+    break;
+    }
+
+    return status;
```
See the comment about `#define HIP_BLAS_STATUS_CHECK(cmd)`

CAHEK7 commented on src/gemm_v2.cpp:
```diff
@@ -613,6 +790,21 @@ miopenStatus_t CallGemm(const Handle& handle,
+           hipblasStatus_t::HIPBLAS_STATUS_SUCCESS)
+        {
+            MIOPEN_THROW(miopenStatusInternalError, "hipBlasLt error encountered");
+        }
```
See the comment about `#define HIP_BLAS_STATUS_CHECK(cmd)`

CAHEK7 commented on src/gemm_v2.cpp:
```diff
@@ -893,6 +1085,21 @@ miopenStatus_t CallGemmStridedBatched(const Handle& handle,
+           hipblasStatus_t::HIPBLAS_STATUS_SUCCESS)
+        {
+            MIOPEN_THROW(miopenStatusInternalError, "hipBlasLt error encountered");
+        }
```
See the comment about `#define HIP_BLAS_STATUS_CHECK(cmd)`

CAHEK7 commented on src/gemm_v2.cpp:
```diff
@@ -1170,6 +1377,21 @@ miopenStatus_t CallGemmStridedBatchedSequential(const Handle& handle,
+           hipblasStatus_t::HIPBLAS_STATUS_SUCCESS)
+        {
+            MIOPEN_THROW(miopenStatusInternalError, "hipBlasLt error encountered");
+        }
```
See the comment about `#define HIP_BLAS_STATUS_CHECK(cmd)`

CAHEK7 commented on src/gemm_v2.cpp:
```diff
@@ -613,6 +790,21 @@ miopenStatus_t CallGemm(const Handle& handle,
+    case GemmBackend_t::hipblaslt: {
+#if MIOPEN_BACKEND_HIP
+        MIOPEN_LOG_FUNCTION("hipBLASLt");
+        HipEventProfiler profiler(handle);
```
I think it would be better to put it as close to the kernel as possible, otherwise any host-related delays may affect the measurements.
Also it can be placed at the lowest level function and removed from multiple caller functions.

CAHEK7 commented on src/gemm_v2.cpp:
```diff
@@ -893,6 +1085,21 @@ miopenStatus_t CallGemmStridedBatched(const Handle& handle,
+    case GemmBackend_t::hipblaslt: {
+#if MIOPEN_BACKEND_HIP
+        MIOPEN_LOG_FUNCTION("hipBLASLt");
+        HipEventProfiler profiler(handle);
```
It can be removed.

CAHEK7 commented on src/gemm_v2.cpp:
```diff
@@ -1170,6 +1377,21 @@ miopenStatus_t CallGemmStridedBatchedSequential(const Handle& handle,
+    case GemmBackend_t::hipblaslt: {
+#if MIOPEN_BACKEND_HIP
+        MIOPEN_LOG_FUNCTION("hipBLASLt");
+        HipEventProfiler profiler(handle);
```
It can be removed.

CAHEK7 commented on src/hipoc/hipoc_kernel.cpp:
```diff
@@ -42,6 +43,29 @@ MIOPEN_DECLARE_ENV_VAR_STR(MIOPEN_DEVICE_ARCH)
 namespace miopen {
 
+HipEventProfiler::HipEventProfiler(const Handle& handle_)
+    : handle(handle_), event_time(0.0f), start(nullptr), stop(nullptr)
```
Since you are improving this class, I would strongly suggest to remove `event_time` from here.
This member can't be used in any case, since it is computed in the destructor., so it can be just a local variable.

BrianHarrisonAMD replied:
Agreed!

CAHEK7 commented on src/include/miopen/hipoc_kernel.hpp:
```diff
@@ -46,6 +46,17 @@ inline HipEventPtr make_hip_event()
+struct HipEventProfiler
+{
+    const Handle& handle;
+    float event_time;
```
It can be removed.

CAHEK7 reviewed (changes requested):
Despite the comments, I have two more important things to mention:
1) some basic tests - just few sanity checks of all the wrappers.
2) some perf measurements - it can be slower than rocblas backend and switching it by defaults may affect the whole rnn performance.

shurale-nkn reviewed (changes requested):
is it guaranteed that the hipBLASLt is always installed in the system with HIP?

BrianHarrisonAMD commented:
Thanks for the feedback @CAHEK7 & @shurale-nkn !

BrianHarrisonAMD committed:
Fix clang tidy error

BrianHarrisonAMD committed:
Update hipblaslt calls to throw instead of checking status, and remove unused HipEventProfiler member variable

BrianHarrisonAMD committed:
Apply formatting and throw if handle creation fails

BrianHarrisonAMD committed:
Add exception when using hipBLASLt with deterministic mode enabled, and update conditional check for RNNs

BrianHarrisonAMD committed:
Remove conditional compilation checks for MIOPEN_BACKEND_HIP

apwojcik commented on CMakeLists.txt:
```diff
@@ -336,6 +336,9 @@ if( MIOPEN_BACKEND STREQUAL "HIP" OR MIOPEN_BACKEND STREQUAL "HIPOC" OR MIOPEN_B
     endif()
+
+    # hipblaslt
+    find_package(hipblaslt REQUIRED PATHS /opt/rocm)
```
Please make it `find_package(hipblaslt REQUIRED PATHS /opt/rocm $ENV{HIP_PATH})` to accommodate Windows.

apwojcik commented on src/CMakeLists.txt:
```diff
@@ -858,6 +853,10 @@ if(rocblas_FOUND)
 endif()
 
+target_link_libraries( MIOpen INTERFACE $<BUILD_INTERFACE:hipblaslt> )
+target_link_libraries( MIOpen PRIVATE hipblaslt )
```
That is not portable between OSes. Please use the imported target `roc::hipblaslt` instead of the library name.

junliume committed:
fix clang format issue

CAHEK7 commented:
May I also ask to update https://github.com/ROCm/MIOpen/blob/e874eb2b8618092a29b77ca5feb737ab29b81e92/src/solver/mha/mha_solver_backward.cpp#L93 and https://github.com/ROCm/MIOpen/blob/e874eb2b8618092a29b77ca5feb737ab29b81e92/src/solver/mha/mha_solver_forward.cpp#L80
and replace `MIOPEN_USE_GEMM` with `MIOPEN_USE_ROCBLAS`.
I don't think that it is super critical, but since both algorithms are using rocblas explicitly, it's better to reflect it.

BrianHarrisonAMD committed:
Merge remote-tracking branch 'origin/develop' into bharriso/enable-hipblaslt-backend

BrianHarrisonAMD committed:
Fix tidy error

BrianHarrisonAMD committed:
Fix windows build issues, and clean-up changes from feedback

iq136boy commented on src/include/miopen/solver/implicitgemm_ck_util.hpp:
```diff
@@ -111,27 +111,7 @@ bool IsCKApplicable(const ProblemDescriptionType& problem)
-        handle.AccumKernelTime(event_time);
-    }
-};
+using WorkAroundHipEventProfiler = HipEventProfiler;
```
LGTM

BrianHarrisonAMD committed:
Fix tidy errors

apwojcik commented:
@junliume @BrianHarrisonAMD, can we make it optional like rocBLAS? Please make it optional with MIOPEN_USE_HIPBLASLT. Unfortunately, the HIP SDK package does not include `hipblastlt` on Windows.

BrianHarrisonAMD commented:
> @junliume @BrianHarrisonAMD, can we make it optional like rocBLAS? Please make it optional with MIOPEN_USE_HIPBLASLT. Unfortunately, the HIP SDK package does not include `hipblastlt` on Windows. Alternatively, make it `hipblas` on Windows.

@apwojcik I don't believe the API is the same between hipBLAS & hipBLASLt. Especially since we are using the c++ extension API. I think we would need to add the define, and add conditional compilation checks around everywhere it's used in the source.

We previously discussed having the define for hipBLASLt, and decided against it, but now that we know it won't work for Windows without it, I imagine that changes the decision. @JehandadKhan thoughts?

BrianHarrisonAMD committed:
Remove deterministic checks

BrianHarrisonAMD committed:
Apply formatting

BrianHarrisonAMD committed:
Merge remote-tracking branch 'origin/develop' into bharriso/enable-hipblaslt-backend

CAHEK7 commented:
> @junliume @BrianHarrisonAMD, can we make it optional like rocBLAS? Please make it optional with MIOPEN_USE_HIPBLASLT. Unfortunately, the HIP SDK package does not include `hipblastlt` on Windows.

Just for the clarification - `hipblas` and `hipblaslt` are separated packages which are not related to each other at all. `hipblas` depends on `rocblas` and it's just a wrapper, so it will not appear without it. `hipblaslt` is an independent standalone library.

BrianHarrisonAMD committed:
Add MIOPEN_USE_HIPBLASLT variable for CMake

BrianHarrisonAMD committed:
Remove static linking for hipBLASLt

BrianHarrisonAMD committed:
Add hipblaslt to gtest

BrianHarrisonAMD committed:
Add linking of hipblaslt to gtest executable

BrianHarrisonAMD committed:
Add define and conditional compilation for hipBLASLt

BrianHarrisonAMD committed:
Fix compiler warnings

BrianHarrisonAMD committed:
Add back ${CMAKE_DL_LIBS] that was missing

atamazov commented:
@BrianHarrisonAMD 

> We previously discussed having the define for hipBLASLt, and decided against it, but now that we know it won't work for Windows without it, I imagine that changes the decision. @JehandadKhan thoughts?

There could be various reasons to build MIOpen without hipBlasLt (including Windows), so I believe we need the CMake variable and corresponding macro for C++ sources.

(this is continuation of https://github.com/ROCm/MIOpen/pull/3030#discussion_r1630384883)

atamazov commented on CMakeLists.txt:
```diff
@@ -337,6 +337,19 @@ if( MIOPEN_BACKEND STREQUAL "HIP" OR MIOPEN_BACKEND STREQUAL "HIPOC" OR MIOPEN_B
+        set(MIOPEN_USE_HIPBLASLT OFF CACHE BOOL "")
+    else()
+        set(MIOPEN_USE_HIPBLASLT ON CACHE BOOL "")
+    endif()
```
```suggestion
set_var_to_condition(MIOPEN_USE_HIPBLASLT_DEFAULT NOT WIN32)
option(MIOPEN_USE_HIPBLASLT "Use hipBlasLt" ${MIOPEN_USE_HIPBLASLT_DEFAULT})
```

atamazov commented on CMakeLists.txt:
```diff
@@ -337,6 +337,19 @@ if( MIOPEN_BACKEND STREQUAL "HIP" OR MIOPEN_BACKEND STREQUAL "HIPOC" OR MIOPEN_B
+    endif()
+    if(MIOPEN_USE_HIPBLASLT)
+        find_package(hipblaslt REQUIRED PATHS /opt/rocm $ENV{HIP_PATH})
+        message(STATUS "Build with hipbBLASLt")
```
```suggestion
        message(STATUS "Build with hipbBLASLt ${hipblaslt_VERSION} ${hipblaslt_DIR}")
```

atamazov commented on src/gemm_v2.cpp:
```diff
@@ -30,9 +30,11 @@
+
+#if MIOPEN_USE_HIPBLASLT
+#include <hipblaslt/hipblaslt.h>
+#include <hipblaslt/hipblaslt-ext.hpp>
```
```suggestion
#include <hipblaslt/hipblaslt-ext.hpp>

#define WORKAROUND_HIPBLASLT_ISSUE_832 1 // Later we can define it based on hipblaslt version.
```

BrianHarrisonAMD replied:
Due to a symbol error, I actually had to swap to the c api instead, and now the const cast can be removed.

atamazov commented on src/gemm_v2.cpp:
```diff
@@ -324,41 +325,216 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+    // NOLINTBEGIN(cppcoreguidelines-pro-type-const-cast)
+    inputs.a = const_cast<DataTypeAB*>(static_cast<const DataTypeAB*>(A) + a_offset);
+    inputs.b = const_cast<DataTypeAB*>(static_cast<const DataTypeAB*>(B) + b_offset);
+    // NOLINTEND(cppcoreguidelines-pro-type-const-cast)
```
https://github.com/ROCm/hipBLASLt/issues/832 opened.
```suggestion
#if WORKAROUND_HIPBLASLT_ISSUE_832
    // Note: Need to const cast here due to hipblaslt_ext::GemmInputs API requirements.
    // NOLINTBEGIN(cppcoreguidelines-pro-type-const-cast)
    inputs.a = const_cast<DataTypeAB*>(static_cast<const DataTypeAB*>(A) + a_offset);
    inputs.b = const_cast<DataTypeAB*>(static_cast<const DataTypeAB*>(B) + b_offset);
    // NOLINTEND(cppcoreguidelines-pro-type-const-cast)
#else
    inputs.a = static_cast<const DataTypeAB*>(A) + a_offset;
    inputs.b = static_cast<const DataTypeAB*>(B) + b_offset;
#endif
```

atamazov commented on src/gemm_v2.cpp:
```diff
@@ -324,41 +325,216 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+    inputs.b = const_cast<DataTypeAB*>(static_cast<const DataTypeAB*>(B) + b_offset);
+    // NOLINTEND(cppcoreguidelines-pro-type-const-cast)
+    inputs.c     = static_cast<DataTypeC*>(C) + c_offset;
+    inputs.d     = static_cast<DataTypeC*>(C) + c_offset;
```
[Q] Is it correct?
```suggestion
    inputs.d     = inputs.c;
```

BrianHarrisonAMD replied:
I believe so, it matches how rocBLAS is providing d in the same file:

`            rb_status = miopen_rocblas_gemm_ex(
                handle,
                gemm_desc,
                gemm_desc.transA ? rocblas_operation_transpose : rocblas_operation_none,
                gemm_desc.transB ? rocblas_operation_transpose : rocblas_operation_none,
                gemm_desc.m,
                gemm_desc.n,
                gemm_desc.k,
                &alpha,
                static_cast<const float*>(A) + a_offset,
                rocblas_datatype::rocblas_datatype_f32_r,
                gemm_desc.lda,
                static_cast<const float*>(B) + b_offset,
                rocblas_datatype::rocblas_datatype_f32_r,
                gemm_desc.ldb,
                &beta,
                static_cast<const float*>(C) + c_offset,
                rocblas_datatype::rocblas_datatype_f32_r,
                gemm_desc.ldc,
                static_cast<float*>(C) + c_offset,
                rocblas_datatype::rocblas_datatype_f32_r,
                gemm_desc.ldc,
                rocBlasComputeType(gemm_desc), // rocblas_datatype::rocblas_datatype_f32_r,
                rocblas_gemm_algo::rocblas_gemm_algo_standard,
                0,
                0);`

My understanding is that c == d for how we use this method.

BrianHarrisonAMD replied:
Edit: Ah I see, yes I will change that.

atamazov commented on src/gemm_v2.cpp:
```diff
@@ -324,41 +325,216 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+    inputs.alpha = &alpha;
+    inputs.beta  = &beta;
+
+    // TODO - bharriso - Need to pass down workspace size & pointer.
```
[R] I would use the following format:
```C++
/// \todo <...stuff...> --<githubUserId> <date>
```
If a user does not have a githubUserId, then provide his/her real name or email.

```suggestion
    /// \todo Need to pass down workspace size & pointer. --bharriso June 2024
```

atamazov commented on src/gemm_v2.cpp:
```diff
@@ -324,41 +325,216 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+    inputs.alpha = &alpha;
+    inputs.beta  = &beta;
+
+    // TODO - bharriso - Need to pass down workspace size & pointer.
```
Please also explain why this is needed (in C++ comments). Thanks.

BrianHarrisonAMD replied:
This will be removed before finishing the PR.
Sorry, just posted this ahead of time for general feedback as I was still finishing the final parts.

atamazov commented on src/gemm_v2.cpp:
```diff
@@ -324,41 +325,216 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+    {
+    case miopenInt8: {
+        MIOPEN_THROW(miopenStatusBadParm,
+                     "miopenInt8 type not supported for hipBlasLt GemmBackend");
```
```suggestion
                     "the miopenInt8 type is not supported for hipBlasLt");
```
or simply
```C++
                     "miopenInt8 is not supported for hipBlasLt");
```
Ditto all other places.

atamazov commented on src/gemm_v2.cpp:
```diff
@@ -324,41 +325,216 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+    switch(gemm_desc.dataType)
+    {
+    case miopenInt8: {
+        MIOPEN_THROW(miopenStatusBadParm,
```
This is internal error of the library. BadParm is about incorrect parameter supplied by the user. Ditto below.

BrianHarrisonAMD replied:
Should I also update the rocBLAS errors that are similar?
I was following what was there before, but I agree it looks like it should be swapped.

atamazov replied:
@BrianHarrisonAMD It would be nice, but it's up to you. Thanks.

atamazov commented on src/gemm_v2.cpp:
```diff
@@ -324,41 +325,216 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
-#else
-    gemm_backend_enforced = GemmBackend_t::nogemmbackend;
+#if MIOPEN_USE_HIPBLASLT
+    case 5: gemm_backend_env = GemmBackend_t::hipblaslt; break;
```
Please also update documentation: https://github.com/ROCm/MIOpen/blob/e5b984a5cc39e07aef1d75246d3272d8066e7a91/docs/how-to/debug-log.rst#rocblas-logging-and-behavior

For example `rocBlas logging and behavior` should be renamed to `GEMM backend logging and behavior` etc

BrianHarrisonAMD replied:
Will do! 

Still need to add a decent amount of documentation actually.

atamazov commented on src/gemm_v2.cpp:
```diff
@@ -324,41 +325,216 @@ inline void ProfilingRecordStop(const Handle& handle, HipEventPtr& start, HipEve
+                                       Data_t C,
+                                       std::size_t c_offset)
+{
+    MIOPEN_LOG_FUNCTION("hipBLASLt");
```
MIOPEN_LOG_FUNCTION is intended to be used **_only_** in API layers. It outputs not only function name but also values of its parameters, and combines logging to console and code annotation for ROC-tracer. Example of use:
https://github.com/ROCm/MIOpen/blob/e5b984a5cc39e07aef1d75246d3272d8066e7a91/src/rnn_api.cpp#L388-L398

MIOPEN_LOG_FUNCTION should not be used in this file.

If you need some indicator that a function is called, please use Info2 logging:
```C++
MIOPEN_LOGI2("");
```
Info2 logging is enabled by `export MIOPEN_LOG_LEVEL=6` or higher.

BrianHarrisonAMD replied:
Should I also remove using this from the equivalent rocBLAS calls?
Or leave it for now?

atamazov replied:
@BrianHarrisonAMD This is not necessary, but I would appreciate that. Thanks.

atamazov reviewed (changes requested):
Part 1. Generally, good, but some changes are still required.

BrianHarrisonAMD committed:
Apply formatting

BrianHarrisonAMD committed:
Implement changes from feedback

- Update throw status & message
- Add workaround define for const_cast
- Remove duplicate static_cast

BrianHarrisonAMD committed:
Update CMakeLists.txt variable define for windows

BrianHarrisonAMD committed:
Update gemm_v2 implementation to use hipblaslt C api

BrianHarrisonAMD committed:
Default to preferred backend when override isnt set

BrianHarrisonAMD committed:
Merge remote-tracking branch 'origin/develop' into bharriso/enable-hipblaslt-backend

BrianHarrisonAMD committed:
Add workspace calculation to RNNs for hipBLASLt GEMM

BrianHarrisonAMD committed:
Remove cout statement

BrianHarrisonAMD committed:
Add RAII style memory cleanup for hipBLASLt handles

BrianHarrisonAMD committed:
Fix memory error for transposed matrices

BrianHarrisonAMD committed:
Apply formatting

IMbackK commented:
In the same vein as https://github.com/pytorch/pytorch/pull/128753#discussion_r1649620265 merging this pr and building MIOpen with hipblaslt enabled will silently (for your ci due runtime assertions being disabled there) break miopen on all targets besides gfx90a,gfx94x and gfx11, as miopen being linked to hipblaslt will cause the runtime to assert and or not finish loading its code objects

amberhassaan commented:
> In the same vein as [pytorch/pytorch#128753 (comment)](https://github.com/pytorch/pytorch/pull/128753#discussion_r1649620265) merging this pr and building MIOpen with hipblaslt enabled will silently (for your ci due runtime assertions being disabled there) break miopen on all targets besides gfx90a,gfx94x and gfx11, as miopen being linked to hipblaslt will cause the runtime to assert and or not finish loading its code objects

@IMbackK : I'm not following the logic fully here (I think). Why is `hipBlasLt` trying to call `ExtractFatBinary`? Isn't it the job of `hip-runtime` or `CLR` to deal with code-objects? For example `MIOpen` uses `hipModule` API to load the code-object. What is `hipBlasLt` doing differently that causes these failures?

IMbackK commented:
Hipblaslt dose not directly call ExtractFatBinary.

My understanding of why hipblaslt is different comes from looking at the backtrace starting at the clr assert and expiramentng with linking to hipblaslt in a hip test sample (and then not useing), so it may not be complete, anyhow iiuc:

For eatch object you dlopen/is loaded by ldd that contains hip code, clr will run ExtractFatBinary to get the code objects for all the targets. If you look at clr source it becomes clear that it here makes the assumption that code objects for all available gpu targets in the system must be available, and it asserts when it dosent find one for any gpu in the system.

hipblaslt is special here because if you look at it with roc-obj-ls it 1. contains gpu code in the core library not just its modules and 2. contains code objects only for a very narrow set of architectures.

the problem is that in your images uses for the basis of the ci of various rocm components assertions where disabled (terrible idea), this causes clr to not assert when it cant find a code object for an architecture it wants, instead the loading of the code objects simply returns when the unsupported combination is encountered. What happens next in this case depends on the system configuration.

1. if all gpus are supported by hipblaslt, everything is fine
2. if none of the gpus are supported clr will return without having loaded anything, calling any hipblaslt function that launches a kernel will cause a crash in clr, but in this case you dont expect it to work any way
3. if there are hipblaslt supported and unsupported gpus clr, depending on the order in which amdgpu.ko initalized the gpus, will return without haveing loaded any code objects for hipblaslt, any use of its functions that launch a kernel will again cause a crash regardless of the active gpu.

amberhassaan commented:
@IMbackK : Thank you for the detailed comment. It makes sense now. It seems like we need to open a ticket on `hipBlasLt` to separate out their `gpu` binaries from their `shared lib` and adopt some kind of on-demand loading, e.g., via `hipModuleLoad`. 
CC: @JehandadKhan , @junliume

BrianHarrisonAMD commented:
@IMbackK Do you have a reproducer, or details about the environment set up for this issue?
I raised the issue with the hipblaslt team, and extra details for how to reproduce would be appreciated.

Thanks!

BrianHarrisonAMD committed:
Revert "Add workspace calculation to RNNs for hipBLASLt GEMM"

This reverts commit 3cc1bee7b79b11da705ac32d5440d78b0ad1aabb.

BrianHarrisonAMD committed:
Merge remote-tracking branch 'origin/develop' into bharriso/enable-hipblaslt-backend

BrianHarrisonAMD committed:
Apply formatting

IMbackK commented:
> @IMbackK Do you have a reproducer, or details about the environment set up for this issue? I raised the issue with the hipblaslt team, and extra details for how to reproduce would be appreciated.
> 
> Thanks!

sure so i am running self-build rocm 6.0.2 on arch

repducing this issue is trivial i just added -lhipblaslt here: https://github.com/ROCm/HIP-Examples/blob/cdf9d101acd9a3fc89ee750f73c1f1958cbd5cc3/add4/Makefile#L16 but linking any hip application that dosent use hipblaslt to hipblaslt will do.

Of course unless you have a configuration noted in my comment above and actually try to use hipblaslt on a supported gpu you will not notice that the runtime has entered a failed state unless you also enable assertions in clr, this is achieved by building clr without -DNDEBUG as is default but unlike how it is built for your ci.

the result is when executed on gfx908 (or any arch besides those supported by hipblaslt):

`a.out: /usr/src/debug/hip-runtime-amd/clr-rocm-6.0.0/hipamd/src/hip_code_object.cpp:762: hip::FatBinaryInfo** hip::StatCO::addFatBinary(const void*, bool): Assertion `err == hipSuccess' failed.`

also note that the mere presence of gfx908 (or any gpu besides those supported by hipblaslt) is enough it dose not need to be the one in use.

BrianHarrisonAMD committed:
Add tests for HipBLASLt CallGemm

BrianHarrisonAMD committed:
Apply formatting

BrianHarrisonAMD committed:
Fix exception error codes, swap logging methods, and change define check for mha

BrianHarrisonAMD committed:
Apply formatting

BrianHarrisonAMD committed:
Update documentation

BrianHarrisonAMD committed:
Add exception check tests

BrianHarrisonAMD committed:
Apply formatting

BrianHarrisonAMD committed:
Move CPU Gemm calculation into a utility header

BrianHarrisonAMD committed:
Update gemm_cpu_util to use existing rnn_util for cpu reference gemm.

BrianHarrisonAMD committed:
Add missing white space

BrianHarrisonAMD committed:
Fix naming to be consistent

BrianHarrisonAMD commented on test/gtest/hipblaslt_gemm.cpp:
```diff
@@ -0,0 +1,384 @@
+                         HipBLASLtGEMMTestBFloat16,
+                         testing::ValuesIn(GetTestCases()));
+
+TEST_F(HipBLASLtGEMMTestFloat8, CheckHipBLASLtGEMMException)
+{
+    using e_mask = enabled<Gpu::gfx94X>;
+    using d_mask =
+        disabled<Gpu::gfx103X, Gpu::gfx900, Gpu::gfx906, Gpu::gfx908, Gpu::gfx90A, Gpu::gfx110X>;
+    CheckExceptionsWithSkip<float8, d_mask, e_mask>(miopenDataType_t::miopenFloat8);
+};
+TEST_P(HipBLASLtGEMMTestFloat8, RunHipBLASLtGEMM)
+{
+    using e_mask = enabled<Gpu::gfx94X>;
+    using d_mask =
+        disabled<Gpu::gfx103X, Gpu::gfx900, Gpu::gfx906, Gpu::gfx908, Gpu::gfx90A, Gpu::gfx110X>;
+    RunGemmDescriptors<float8, d_mask, e_mask>(GetParam(), miopenDataType_t::miopenFloat8);
+};
+INSTANTIATE_TEST_SUITE_P(HipBLASLtGEMMTestSet,
+                         HipBLASLtGEMMTestFloat8,
+                         testing::ValuesIn(GetTestCases()));
+
+TEST_F(HipBLASLtGEMMTestBFloat8, CheckHipBLASLtGEMMException)
+{
+    using e_mask = enabled<Gpu::gfx94X>;
+    using d_mask =
+        disabled<Gpu::gfx103X, Gpu::gfx900, Gpu::gfx906, Gpu::gfx908, Gpu::gfx90A, Gpu::gfx110X>;
+    CheckExceptionsWithSkip<bfloat8, d_mask, e_mask>(miopenDataType_t::miopenBFloat8);
+};
+TEST_P(HipBLASLtGEMMTestBFloat8, RunHipBLASLtGEMM)
+{
+    using e_mask = enabled<Gpu::gfx94X>;
+    using d_mask =
+        disabled<Gpu::gfx103X, Gpu::gfx900, Gpu::gfx906, Gpu::gfx908, Gpu::gfx90A, Gpu::gfx110X>;
+    RunGemmDescriptors<bfloat8, d_mask, e_mask>(GetParam(), miopenDataType_t::miopenBFloat8);
+};
+INSTANTIATE_TEST_SUITE_P(HipBLASLtGEMMTestSet,
+                         HipBLASLtGEMMTestBFloat8,
+                         testing::ValuesIn(GetTestCases()));
```
Failing to find a heuristic result for FP8 and BF8 in some cases.
Need to figure out this issue.

BrianHarrisonAMD replied:
Increasing the m, n, & k value fixed the heuristic failing to find for FP8, but still unsure about BF8 failures.

BrianHarrisonAMD replied:
BF8 failures are fixed on newer versions of hipBLASLt.

BrianHarrisonAMD replied:
Fixed by only enabling BF8 when compiling against newer versions of hipBLASLt

BrianHarrisonAMD committed:
Fix FP8 tests failing to find heuristic

apwojcik reviewed (changes requested):
@BrianHarrisonAMD the PR does not compile on Windows

BrianHarrisonAMD committed:
Fix build issue when MIOPEN_USE_HIPBLASLT is off

BrianHarrisonAMD commented:
> @BrianHarrisonAMD the PR does not compile on Windows

Should be fixed now!

apwojcik commented:
> > @BrianHarrisonAMD the PR does not compile on Windows
> 
> Should be fixed now!

It is still failing, but we have a fix for that already (not yet merged): https://github.com/ROCm/MIOpen/pull/3090

BrianHarrisonAMD committed:
Add selectively enabling BF8 support

BrianHarrisonAMD committed:
Fix build issue when MIOPEN_USE_HIPBLASLT is off with nogpu

BrianHarrisonAMD committed:
Add hipBLAS package to CMake

BrianHarrisonAMD committed:
Apply formatting

BrianHarrisonAMD committed:
Update documentation to include hipBLAS dependency

BrianHarrisonAMD committed:
Add missing hipBLAS doc link for install.rst

BrianHarrisonAMD committed:
Merge remote-tracking branch 'origin/develop' into bharriso/enable-hipblaslt-backend

BrianHarrisonAMD committed:
Merge remote-tracking branch 'origin/develop' into bharriso/enable-hipblaslt-backend

BrianHarrisonAMD commented:
Greetings @shurale-nkn!

If you get a chance can you check this PR again?
I believe all the changes you had requested are implemented, and the change requested state is blocking merging.

Thanks!

shurale-nkn commented on src/ocl/rnnocl.cpp:
```diff
@@ -87,6 +87,20 @@ void checkGemmStatusAndLog(miopenStatus_t gemm_status)
+    // Only use the hipblaslt backend when device is MI300, and the datatype is half.
+    // Otherwise, default to rocblas.
+    // Note: environment variable can be used to force a specific backend.
+    return handle.GetDeviceName() == "gfx942" && dataType == miopenDataType_t::miopenHalf
```
I forbid making this change. Only optional by ENV switch, default is rocblas.
I don't see any significant improvements in the local tests.
Moreover, in multithreaded mode current develop shows better performance.
Develop 19.802ms , this PR: 22.89ms. the new value is 15% slower
./MIOpenDriver rnnfp16 -n 1024 -W 224 -H 1000 -l 8 -b 1 -m lstm -p 0 -r 0 -k 32 -F 1 -c 0 -t 1 -w 1 -V 0 --iter 10

BrianHarrisonAMD replied:
Curious why we are getting different results.
When we ran our tests there was a minor improvement using hipBLASLt over rocBLAS 3-4% it seemed on gfx942. 
I would be fine with leaving it only accessible behind the environment variable for now though.

BrianHarrisonAMD replied:
@shurale-nkn oh, I think it could possibly be due to the large performance hit the first time you use hipBLASLt in an application. Finding the heuristic has a long load time the first time running it, but the second run is much quicker, and that first load is ignored in the benchmarks we were running.

Edit: Nope it skips the first run results when timing with --iter 10. 
Only thing I can think of is a different version of hipBLASLt being used.

BrianHarrisonAMD replied:
Re-ran the above to check for me, and I get the following with default rocBLAS & hipBLASLt installed:

**Branch hipBLASLt:**
MIOPEN_GEMM_ENFORCE_BACKEND=5 ./bin/MIOpenDriver rnnfp16 -n 1024 -W 224 -H 1000 -l 8 -b 1 -m lstm -p 0 -r 0 -k 32 -F 1 -c 0 -t 1 -w 1 -V 0 --iter 10
MIOpenDriver rnnfp16 -n 1024 -W 224 -H 1000 -l 8 -b 1 -m lstm -p 0 -r 0 -k 32 -F 1 -c 0 -t 1 -w 1 -V 0 --iter 10
length of data sequence == 1 is short than time sequence == 32, padding the rest of data sequence with 1024
length of data sequence == 1 is short than time sequence == 32, padding the rest of data sequence with 1024
PRNG seed: 12345678
GPU Kernel Time Forward RNN Elapsed: 25.389431 ms
Wall-clock Time Forward RNN Elapsed: 25.442097 ms

**Branch rocBLAS:**
MIOPEN_GEMM_ENFORCE_BACKEND=1 ./bin/MIOpenDriver rnnfp16 -n 1024 -W 224 -H 1000 -l 
8 -b 1 -m lstm -p 0 -r 0 -k 32 -F 1 -c 0 -t 1 -w 1 -V 0 --iter 10
MIOpenDriver rnnfp16 -n 1024 -W 224 -H 1000 -l 8 -b 1 -m lstm -p 0 -r 0 -k 32 -F 1 -c 0 -t 1 -w 1 -V 0 --iter 10
length of data sequence == 1 is short than time sequence == 32, padding the rest of data sequence with 1024
length of data sequence == 1 is short than time sequence == 32, padding the rest of data sequence with 1024
PRNG seed: 12345678
GPU Kernel Time Forward RNN Elapsed: 31.276720 ms
Wall-clock Time Forward RNN Elapsed: 31.345469 ms

**Develop rocBLAS**:
MIOPEN_GEMM_ENFORCE_BACKEND=1 ./bin/MIOpenDriver rnnfp16 -n 1024 -W 224 -H 1000 -l 8 -b 1 -m lstm -p 0 -r 0 -k 32 -F 1 -c 0 -t 1 -w 1 -V 0 --iter 10
MIOpenDriver rnnfp16 -n 1024 -W 224 -H 1000 -l 8 -b 1 -m lstm -p 0 -r 0 -k 32 -F 1 -c 0 -t 1 -w 1 -V 0 --iter 10
length of data sequence == 1 is short than time sequence == 32, padding the rest of data sequence with 1024
length of data sequence == 1 is short than time sequence == 32, padding the rest of data sequence with 1024
PRNG seed: 12345678
GPU Kernel Time Forward RNN Elapsed: 29.586761 ms
Wall-clock Time Forward RNN Elapsed: 29.644012 ms

This is with the following versions:
- hipBLASLt 0.7.0 Tweak 1549b021
- rocBLAS 4.1.0 Tweak cefa4a9b-dirty

shurale-nkn replied:
I'm using tuned rocblas from develop
![image](https://github.com/ROCm/MIOpen/assets/37591772/ed33b3b7-4a92-427a-b79e-7b0216ebff4d)

BrianHarrisonAMD replied:
Ah that explains it.
Ill double check with latest hipBLASLt, but I will swap to make it only usable through the environment variables.

BrianHarrisonAMD replied:
Well wait, when I tested it originally, it was with a tuned version of rocBLAS:

#define ROCBLAS_VERSION_MAJOR       4
#define ROCBLAS_VERSION_MINOR       3
#define ROCBLAS_VERSION_PATCH       0
#define ROCBLAS_VERSION_TWEAK       18feb608
#define ROCBLAS_TENSILE_COMMIT_ID   "18feb6085f70551903db177797a90d93b88e688b" , "1e5f2423ddf5594077e883da138f7efaf84f77bd"

Let me check again with this and latest hipBLASLt.

BrianHarrisonAMD replied:
Even with the newer version of rocBLAS I am still getting similar results:

MIOPEN_GEMM_ENFORCE_BACKEND=1 ./bin/MIOpenDriver rnnfp16 -n 1024 -W 224 -H 1000 -l 8 -b 1 -m lstm -p 0 -r 0 -k 32 -F 1 -c 0 -t 1 -w 1 -V 0 -iter 10
MIOpenDriver rnnfp16 -n 1024 -W 224 -H 1000 -l 8 -b 1 -m lstm -p 0 -r 0 -k 32 -F 1 -c 0 -t 1 -w 1 -V 0 -iter 10
length of data sequence == 1 is short than time sequence == 32, padding the rest of data sequence with 1024
length of data sequence == 1 is short than time sequence == 32, padding the rest of data sequence with 1024
PRNG seed: 12345678
GPU Kernel Time Forward RNN Elapsed: 23.978241 ms
Wall-clock Time Forward RNN Elapsed: 24.052076 ms

They seem to be about the same to me though, and I will put it back to using rocBLAS as the default.

BrianHarrisonAMD replied:
Alright @shurale-nkn, I swapped it so now it should only be usable via the environment variable.

Are you good with these changes?

shurale-nkn commented on src/hip/handlehip.cpp:
```diff
@@ -230,23 +234,43 @@ struct HandleImpl
 #endif
+#if MIOPEN_USE_HIPBLASLT
+    hipblasLt_handle_ptr hip_blasLt_handle;
+    using HipblasLtHandlePtrPool = std::vector<hipblasLt_handle_ptr>;
```
This is useless and wrong. If a hipblasLt_handle is created without connection to a specific stream, there is no point in producing entities

BrianHarrisonAMD replied:
@shurale-nkn I actually asked the hipBLASLt team about this when implementing.

Here is what I asked:

"Previously with rocBLAS, the rocBLAS handle allocated device memory that could not be shared by multiple async streams at the same time, and thus we had to keep track of which rocBLAS handle corresponded to which stream, and make a separate one for each active stream being used.
 
With hipBLASLt, it looks like that is no longer a limitation, and we can have a single hipblasLtHandle that will work for all streams. Is that accurate? Specifically, I don't see that mentioned in the documentation, and the API appears to take in streams instead of having api for setting the stream on the handle."

and here was the response:

"No, we have similar behavior. Different stream should use different handle."

I believe this means we need the above, otherwise if we use hipBLASLt in a multi-threaded situation it won't work properly.

shurale-nkn replied:
I couldn't find any confirmations in the documentation, but looking at the project code, they really allocate something.
It's a strange design decision to make part of the api without a handle, and make an implicit dependency between the handle and the used stream. 
but okay, we'll have to create them in the same manner.

BrianHarrisonAMD replied:
Yea I was surprised as well, and originally had it as just a single handle.

BrianHarrisonAMD committed:
Revert defaulting rnn16 to use hipBLASLt on gfx942

shurale-nkn reviewed (approved):
LGTM.
But I will leave additional notes here for the future. After backend switch host time at rnn increased by about 20%, in some configurations this will be critical. It is necessary to analyze whether we can improve the implementation on our side. miopen_hipblasLt_gemm the function now looks very fat for a simple wrapper that is executed for each GEMM call, many calls for handle configuration.  But perhaps the problem is in the hipblasLt itself, at first glance, multiple transformations are also constantly taking place inside.

BrianHarrisonAMD committed:
Merge remote-tracking branch 'origin/develop' into bharriso/enable-hipblaslt-backend

IMbackK commented:
> Hipblaslt dose not directly call ExtractFatBinary.
> 
> My understanding of why hipblaslt is different comes from looking at the backtrace starting at the clr assert and expiramentng with linking to hipblaslt in a hip test sample (and then not useing), so it may not be complete, anyhow iiuc:
> 
> For eatch object you dlopen/is loaded by ldd that contains hip code, clr will run ExtractFatBinary to get the code objects for all the targets. If you look at clr source it becomes clear that it here makes the assumption that code objects for all available gpu targets in the system must be available, and it asserts when it dosent find one for any gpu in the system.
> 
> hipblaslt is special here because if you look at it with roc-obj-ls it 1. contains gpu code in the core library not just its modules and 2. contains code objects only for a very narrow set of architectures.
> 
> the problem is that in your images uses for the basis of the ci of various rocm components assertions where disabled (terrible idea), this causes clr to not assert when it cant find a code object for an architecture it wants, instead the loading of the code objects simply returns when the unsupported combination is encountered. What happens next in this case depends on the system configuration.
> 
>     1. if all gpus are supported by hipblaslt, everything is fine
> 
>     2. if none of the gpus are supported clr will return without having loaded anything, calling any hipblaslt function that launches a kernel will cause a crash in clr, but in this case you dont expect it to work any way
> 
>     3. if there are hipblaslt supported and unsupported gpus clr, depending on the order in which amdgpu.ko initalized the gpus, will return without haveing loaded any code objects for hipblaslt, any use of its functions that launch a kernel will again cause a crash regardless of the active gpu.

Please dont merge this without addressing this problem, i dont want to have to build yet another rocm component by hand just to undo the damage hipblaslt is doing everywhere.

junliume commented:
> > Hipblaslt dose not directly call ExtractFatBinary.
> > My understanding of why hipblaslt is different comes from looking at the backtrace starting at the clr assert and expiramentng with linking to hipblaslt in a hip test sample (and then not useing), so it may not be complete, anyhow iiuc:
> > For eatch object you dlopen/is loaded by ldd that contains hip code, clr will run ExtractFatBinary to get the code objects for all the targets. If you look at clr source it becomes clear that it here makes the assumption that code objects for all available gpu targets in the system must be available, and it asserts when it dosent find one for any gpu in the system.
> > hipblaslt is special here because if you look at it with roc-obj-ls it 1. contains gpu code in the core library not just its modules and 2. contains code objects only for a very narrow set of architectures.
> > the problem is that in your images uses for the basis of the ci of various rocm components assertions where disabled (terrible idea), this causes clr to not assert when it cant find a code object for an architecture it wants, instead the loading of the code objects simply returns when the unsupported combination is encountered. What happens next in this case depends on the system configuration.
> > ```
> > 1. if all gpus are supported by hipblaslt, everything is fine
> > 
> > 2. if none of the gpus are supported clr will return without having loaded anything, calling any hipblaslt function that launches a kernel will cause a crash in clr, but in this case you dont expect it to work any way
> > 
> > 3. if there are hipblaslt supported and unsupported gpus clr, depending on the order in which amdgpu.ko initalized the gpus, will return without haveing loaded any code objects for hipblaslt, any use of its functions that launch a kernel will again cause a crash regardless of the active gpu.
> > ```
> 
> Please dont merge this without addressing this problem, i dont want to have to build yet another rocm component by hand just to undo the damage hipblaslt is doing everywhere.

@IMbackK this PR conditionally enable it as an optional path which is off by default, please elaborate more on the requests since we need to move swiftly on more feature enablement. Thanks!

BrianHarrisonAMD commented:
@IMbackK I don't think this should cause any issues for you.
It's using rocBLAS by default in all modes now, and only enabled if you opt into the environment variable override.

I suppose you might get asserts, but you can also disable compiling with hipBLASLt support if needed via the variable ``-DMIOPEN_USE_HIPBLASLT=Off``

junliume commented:
@atamazov @amberhassaan @CAHEK7 if no objections, we plan to move forward with this PR soon.

IMbackK commented:
> @IMbackK I don't think this should cause any issues for you. It's using rocBLAS by default in all modes now, and only enabled if you opt into the environment variable override.
> 
> I suppose you might get asserts, but you can also disable compiling with hipBLASLt support if needed via the variable `-DMIOPEN_USE_HIPBLASLT=Off`

This will cause issues for me and and any configuration mentioned since you set miopen to be linked against hipBLASLt by default here https://github.com/ROCm/MIOpen/pull/3030/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR342 
Thus in the rocm offical packages, the docker images etc miopen will cease to work correctly on all non gfx90a/gfx94x/gfx11 targets 
sure your ci will pass this since it has assertions disabled in clr and dose not actually hit the undefined behavior in clr but its still fundamentally broken

Using hipblaslt or not at runtime (chosen by the envar or arbitrary other runtime means) is irrelevant the runtime has failed before the application that uses miopen has even reached main() 

Sure i can recompile miopen with this pr disabled as you mentioned, but i cant see how it can be acceptable to merge this and enable it by default knowing its broken.

IMbackK commented:
Since this is an issue in hipblaslt, merging this is fine as long as it stays disabled (important: AT COMPILETIME to avoid triggering the issue) by default until hipblaslt has been fixed.
